{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7d86a7aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import requests as re\n",
    "import untangle\n",
    "from collections import Counter\n",
    "import json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3e9cd6ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "baseurl = \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/\"\n",
    "esearch = \"esearch.fcgi\"\n",
    "efetch = \"efetch.fcgi\"\n",
    "db = \"protein\"\n",
    "species = []\n",
    "glands = []\n",
    "variants = pd.read_csv(\"spider_silks.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "18fb553b",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"silks.json\", \"r\") as f:\n",
    "    silk_json = json.loads(f.read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "612e73a9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'name': 'Major Ampullate', 'alt': 'dragline'},\n",
       " {'name': 'Flagelliform', 'alt': ''},\n",
       " {'name': 'Aggregate spidroin', 'alt': ''},\n",
       " {'name': 'Minor ampullate', 'alt': ''},\n",
       " {'name': 'Tubuliform', 'alt': ''},\n",
       " {'name': 'Aciniform', 'alt': ''},\n",
       " {'name': 'Pyriform', 'alt': ''},\n",
       " {'name': 'Cribellar', 'alt': ''}]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "silk_json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "28bf12ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_db_request(link, params):\n",
    "    link = baseurl + link\n",
    "    request = re.get(link, params=params)\n",
    "    return request.text\n",
    "\n",
    "def get_search_ids(gland):\n",
    "    query = '\"{}\"'.format(gland[\"name\"].replace(\" \", \"+\"))\n",
    "    params = {\"db\": db, \"term\": query, \"retmax\" : 1500}       \n",
    "    request = make_db_request(esearch, params)\n",
    "    searches = untangle.parse(request)\n",
    "    idlist = [id.cdata for id in searches.eSearchResult.IdList.children]\n",
    "    return idlist\n",
    "\n",
    "def parse_features(features):\n",
    "    for feat in features.children:\n",
    "        if feat.GBFeature_key == \"CDS\":\n",
    "            qualifiers = feat.GBFeature_quals.children\n",
    "            for qual in qualifiers:\n",
    "                if (qual.GBQualifier_name == \"gene\"):\n",
    "                    return qual.GBQualifier_value.cdata\n",
    "    return \"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "15f868c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_refs(reflist):\n",
    "    ref = reflist.GBReference\n",
    "    if (len(ref) >= 1 and ref[0] is not None):\n",
    "        return ref[0].GBReference_title.cdata\n",
    "    else:\n",
    "        return \"\"\n",
    "    \n",
    "def filter_searches(definition, taxonomy, name, alt=\"\"):\n",
    "    definition = definition.lower()\n",
    "    if (\"Araneae\" not in taxonomy):\n",
    "        return False\n",
    "    if ((name.lower() not in definition) and (alt.lower()) not in definition):\n",
    "        return False\n",
    "    return True\n",
    "    \n",
    "def parse_document(xmlstr, gland_obj):\n",
    "    genbankurl = \"https://www.ncbi.nlm.nih.gov/protein/\"\n",
    "    doc = untangle.parse(xmlstr)\n",
    "    gbset =  doc.GBSet\n",
    "    silk_db = []\n",
    "    gland = gland_obj[\"name\"]\n",
    "    alt = gland_obj[\"alt\"]\n",
    "    for gbseq in gbset.children:\n",
    "        definition = gbseq.GBSeq_definition.cdata\n",
    "        taxonomy = gbseq.GBSeq_taxonomy.cdata\n",
    "        if filter_searches(definition, taxonomy, gland, alt):\n",
    "            sequence = gbseq.GBSeq_sequence.cdata.upper()\n",
    "            species = gbseq.GBSeq_organism.cdata\n",
    "            size = gbseq.GBSeq_length.cdata\n",
    "            partial = \"partial\" if \"partial\" in definition else \"\"\n",
    "            accession = gbseq.GBSeq_accession_version.cdata\n",
    "            gene = parse_features(gbseq.GBSeq_feature_table)\n",
    "            database = '=HYPERLINK(\"{url}{acc}\",\"Genbank:{acc}\")'.format(acc=accession, url=genbankurl)\n",
    "            ref = \"\"\n",
    "            if \"GBSeq_references\" in gbseq:\n",
    "                ref = get_refs(gbseq.GBSeq_references)\n",
    "            silk_db += [[gene, gland, species, \"\", size, \"\", database, sequence, partial, \"\", ref]]\n",
    "    return silk_db\n",
    "\n",
    "def get_document(idlist, gland, columns):\n",
    "    retstart = 0\n",
    "    retmax = 200\n",
    "    silk_db = []\n",
    "    while retstart < len(idlist):\n",
    "        ids = \",\".join(idlist[retstart:retstart+retmax])\n",
    "        params = {\"db\": db, \"id\": ids, \"rettype\": \"gp\", \"retmode\": \"xml\"}\n",
    "        xmlstr = make_db_request(efetch, params)\n",
    "        silk_db += parse_document(xmlstr, gland)\n",
    "        retstart += retmax\n",
    "    print(len(silk_db))\n",
    "    return pd.DataFrame(silk_db, columns=columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "ae84f4a7",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'ids' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/hr/5b8zkjbs6d3007k2cnctrgh40000gn/T/ipykernel_1892/4023290486.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mget_document\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mids\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"Major ampullate\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msilk_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'ids' is not defined"
     ]
    }
   ],
   "source": [
    "get_document(ids, \"Major ampullate\", silk_df.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "71d6fa13",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'name': 'Major Ampullate', 'alt': 'dragline'}\n",
      "407\n",
      "{'name': 'Flagelliform', 'alt': ''}\n",
      "58\n",
      "{'name': 'Aggregate spidroin', 'alt': ''}\n",
      "0\n",
      "{'name': 'Minor ampullate', 'alt': ''}\n",
      "148\n",
      "{'name': 'Tubuliform', 'alt': ''}\n",
      "214\n",
      "{'name': 'Aciniform', 'alt': ''}\n",
      "127\n",
      "{'name': 'Pyriform', 'alt': ''}\n",
      "39\n",
      "{'name': 'Cribellar', 'alt': ''}\n",
      "8\n"
     ]
    }
   ],
   "source": [
    "def create_db(df):\n",
    "    for gland in silk_json:\n",
    "        print (gland)\n",
    "        ids = get_search_ids(gland)\n",
    "        df = pd.concat([df, get_document(ids, gland, df.columns)])\n",
    "    return df\n",
    "\n",
    "silk_df = pd.DataFrame(columns=[\"Spidrion\", \"Gland / Silk Type\", \"Species\", \"Spinneret Used\", \"Size\", \"Use\", \"Database(s)\", \"Sequence\", \"Notes\", \"Experiment Considerations\", \"References\"])\n",
    "silk_df = create_db(silk_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "a383ec9f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Spidrion</th>\n",
       "      <th>Gland / Silk Type</th>\n",
       "      <th>Species</th>\n",
       "      <th>Spinneret Used</th>\n",
       "      <th>Size</th>\n",
       "      <th>Use</th>\n",
       "      <th>Database(s)</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Notes</th>\n",
       "      <th>Experiment Considerations</th>\n",
       "      <th>References</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td></td>\n",
       "      <td>Major Ampullate</td>\n",
       "      <td>Caerostris darwini</td>\n",
       "      <td></td>\n",
       "      <td>169</td>\n",
       "      <td></td>\n",
       "      <td>=HYPERLINK(\"https://www.ncbi.nlm.nih.gov/prote...</td>\n",
       "      <td>MVLIIGLTSVTSLKVEELQKSYRIAFNERQKRSPKEDCSNSYNDVL...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MaSp3D1</td>\n",
       "      <td>Major Ampullate</td>\n",
       "      <td>Caerostris darwini</td>\n",
       "      <td></td>\n",
       "      <td>2328</td>\n",
       "      <td></td>\n",
       "      <td>=HYPERLINK(\"https://www.ncbi.nlm.nih.gov/prote...</td>\n",
       "      <td>MGWITNIIIFVLFCTQANAFNKETVMGPGMMRDFMNGMSNAMANSG...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MaSp1F2</td>\n",
       "      <td>Major Ampullate</td>\n",
       "      <td>Caerostris darwini</td>\n",
       "      <td></td>\n",
       "      <td>2985</td>\n",
       "      <td></td>\n",
       "      <td>=HYPERLINK(\"https://www.ncbi.nlm.nih.gov/prote...</td>\n",
       "      <td>MTWTSRLALSLLVAICTQSMFALGQDNTPWSSTGTAESFMSSFMSA...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MaSp2C1</td>\n",
       "      <td>Major Ampullate</td>\n",
       "      <td>Caerostris darwini</td>\n",
       "      <td></td>\n",
       "      <td>710</td>\n",
       "      <td></td>\n",
       "      <td>=HYPERLINK(\"https://www.ncbi.nlm.nih.gov/prote...</td>\n",
       "      <td>GGYGPGGQGPSGPGSQGPGGQGPYGPGGSAAAAAAAGGYGPGGQGP...</td>\n",
       "      <td>partial</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MaSp2C1</td>\n",
       "      <td>Major Ampullate</td>\n",
       "      <td>Caerostris darwini</td>\n",
       "      <td></td>\n",
       "      <td>3227</td>\n",
       "      <td></td>\n",
       "      <td>=HYPERLINK(\"https://www.ncbi.nlm.nih.gov/prote...</td>\n",
       "      <td>MYSSTRLALTLLAVLCTQAVFTAAQAPSPWKRTAPAERFITSFITA...</td>\n",
       "      <td>partial</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CrSp</td>\n",
       "      <td>Cribellar</td>\n",
       "      <td>Octonoba varians</td>\n",
       "      <td></td>\n",
       "      <td>4449</td>\n",
       "      <td></td>\n",
       "      <td>=HYPERLINK(\"https://www.ncbi.nlm.nih.gov/prote...</td>\n",
       "      <td>VVLSRKLKRILPAALLTADLSSGLGRLGAGAPASRYAQLVAGSIAL...</td>\n",
       "      <td>partial</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CrSp</td>\n",
       "      <td>Cribellar</td>\n",
       "      <td>Octonoba sybotides</td>\n",
       "      <td></td>\n",
       "      <td>2426</td>\n",
       "      <td></td>\n",
       "      <td>=HYPERLINK(\"https://www.ncbi.nlm.nih.gov/prote...</td>\n",
       "      <td>MDWLTRSAMLLLLIFFRCNGIFGQSIAGQEISAPSNINSAQNFATS...</td>\n",
       "      <td>partial</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CrSp</td>\n",
       "      <td>Cribellar</td>\n",
       "      <td>Octonoba sybotides</td>\n",
       "      <td></td>\n",
       "      <td>986</td>\n",
       "      <td></td>\n",
       "      <td>=HYPERLINK(\"https://www.ncbi.nlm.nih.gov/prote...</td>\n",
       "      <td>SAAAAAQAVEREEGVGTEGVSAGEGLRDITEGAKKVAQSAEKSEIF...</td>\n",
       "      <td>partial</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CrSp</td>\n",
       "      <td>Cribellar</td>\n",
       "      <td>Octonoba grandiprojecta</td>\n",
       "      <td></td>\n",
       "      <td>1964</td>\n",
       "      <td></td>\n",
       "      <td>=HYPERLINK(\"https://www.ncbi.nlm.nih.gov/prote...</td>\n",
       "      <td>DLTVHLPSAPEHSEKLASAFELASGTSTQVVVPSQTESELSTLTEG...</td>\n",
       "      <td>partial</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CrSp</td>\n",
       "      <td>Cribellar</td>\n",
       "      <td>Octonoba grandiprojecta</td>\n",
       "      <td></td>\n",
       "      <td>643</td>\n",
       "      <td></td>\n",
       "      <td>=HYPERLINK(\"https://www.ncbi.nlm.nih.gov/prote...</td>\n",
       "      <td>EPSAAAEAVEFEEAVAEKGVSAGEGLRDIIVGVEKGVQSAEKSETS...</td>\n",
       "      <td>partial</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1001 rows Ã— 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Spidrion Gland / Silk Type                  Species Spinneret Used  Size  \\\n",
       "0             Major Ampullate       Caerostris darwini                  169   \n",
       "1   MaSp3D1   Major Ampullate       Caerostris darwini                 2328   \n",
       "2   MaSp1F2   Major Ampullate       Caerostris darwini                 2985   \n",
       "3   MaSp2C1   Major Ampullate       Caerostris darwini                  710   \n",
       "4   MaSp2C1   Major Ampullate       Caerostris darwini                 3227   \n",
       "..      ...               ...                      ...            ...   ...   \n",
       "3      CrSp         Cribellar         Octonoba varians                 4449   \n",
       "4      CrSp         Cribellar       Octonoba sybotides                 2426   \n",
       "5      CrSp         Cribellar       Octonoba sybotides                  986   \n",
       "6      CrSp         Cribellar  Octonoba grandiprojecta                 1964   \n",
       "7      CrSp         Cribellar  Octonoba grandiprojecta                  643   \n",
       "\n",
       "   Use                                        Database(s)  \\\n",
       "0       =HYPERLINK(\"https://www.ncbi.nlm.nih.gov/prote...   \n",
       "1       =HYPERLINK(\"https://www.ncbi.nlm.nih.gov/prote...   \n",
       "2       =HYPERLINK(\"https://www.ncbi.nlm.nih.gov/prote...   \n",
       "3       =HYPERLINK(\"https://www.ncbi.nlm.nih.gov/prote...   \n",
       "4       =HYPERLINK(\"https://www.ncbi.nlm.nih.gov/prote...   \n",
       "..  ..                                                ...   \n",
       "3       =HYPERLINK(\"https://www.ncbi.nlm.nih.gov/prote...   \n",
       "4       =HYPERLINK(\"https://www.ncbi.nlm.nih.gov/prote...   \n",
       "5       =HYPERLINK(\"https://www.ncbi.nlm.nih.gov/prote...   \n",
       "6       =HYPERLINK(\"https://www.ncbi.nlm.nih.gov/prote...   \n",
       "7       =HYPERLINK(\"https://www.ncbi.nlm.nih.gov/prote...   \n",
       "\n",
       "                                             Sequence    Notes  \\\n",
       "0   MVLIIGLTSVTSLKVEELQKSYRIAFNERQKRSPKEDCSNSYNDVL...            \n",
       "1   MGWITNIIIFVLFCTQANAFNKETVMGPGMMRDFMNGMSNAMANSG...            \n",
       "2   MTWTSRLALSLLVAICTQSMFALGQDNTPWSSTGTAESFMSSFMSA...            \n",
       "3   GGYGPGGQGPSGPGSQGPGGQGPYGPGGSAAAAAAAGGYGPGGQGP...  partial   \n",
       "4   MYSSTRLALTLLAVLCTQAVFTAAQAPSPWKRTAPAERFITSFITA...  partial   \n",
       "..                                                ...      ...   \n",
       "3   VVLSRKLKRILPAALLTADLSSGLGRLGAGAPASRYAQLVAGSIAL...  partial   \n",
       "4   MDWLTRSAMLLLLIFFRCNGIFGQSIAGQEISAPSNINSAQNFATS...  partial   \n",
       "5   SAAAAAQAVEREEGVGTEGVSAGEGLRDITEGAKKVAQSAEKSEIF...  partial   \n",
       "6   DLTVHLPSAPEHSEKLASAFELASGTSTQVVVPSQTESELSTLTEG...  partial   \n",
       "7   EPSAAAEAVEFEEAVAEKGVSAGEGLRDIIVGVEKGVQSAEKSETS...  partial   \n",
       "\n",
       "   Experiment Considerations References  \n",
       "0                                        \n",
       "1                                        \n",
       "2                                        \n",
       "3                                        \n",
       "4                                        \n",
       "..                       ...        ...  \n",
       "3                                        \n",
       "4                                        \n",
       "5                                        \n",
       "6                                        \n",
       "7                                        \n",
       "\n",
       "[1001 rows x 11 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "silk_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "88d2513a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Agelenopsis aperta',\n",
       " 'Aptostichus sp. AS220',\n",
       " 'Araneus bicentenarius',\n",
       " 'Araneus diadematus',\n",
       " 'Araneus gemmoides',\n",
       " 'Araneus ventricosus',\n",
       " 'Argiope amoena',\n",
       " 'Argiope argentata',\n",
       " 'Argiope aurantia',\n",
       " 'Argiope bruennichi',\n",
       " 'Argiope trifasciata',\n",
       " 'Argyroneta aquatica',\n",
       " 'Bothriocyrtum californicum',\n",
       " 'Caerostris darwini',\n",
       " 'Caerostris extrusa',\n",
       " 'Cybaeus angustiarum',\n",
       " 'Cyrtophora moluccensis',\n",
       " 'Deinopis spinosa',\n",
       " 'Diguetia canities',\n",
       " 'Euagrus chisoseus',\n",
       " 'Euprosthenops australis',\n",
       " 'Euprosthenops sp. NNP-2003',\n",
       " 'Gasteracantha cancriformis',\n",
       " 'Gea heptagon',\n",
       " 'Kukulcania hibernalis',\n",
       " 'Latrodectus geometricus',\n",
       " 'Latrodectus hasseltii',\n",
       " 'Latrodectus hesperus',\n",
       " 'Latrodectus mactans',\n",
       " 'Latrodectus tredecimguttatus',\n",
       " 'Macrothele holsti',\n",
       " 'Metepeira grandiosa',\n",
       " 'Nephila pilipes',\n",
       " 'Nephilengys cruentata',\n",
       " 'Octonoba grandiconcava',\n",
       " 'Octonoba grandiprojecta',\n",
       " 'Octonoba okinawensis',\n",
       " 'Octonoba sybotides',\n",
       " 'Octonoba varians',\n",
       " 'Octonoba yesoensis',\n",
       " 'Oedothorax gibbosus',\n",
       " 'Parasteatoda tepidariorum',\n",
       " 'Parawixia bistriata',\n",
       " 'Peucetia viridans',\n",
       " 'Pholcus phalangioides',\n",
       " 'Physocyclus mexicanus',\n",
       " 'Plectreurys tristis',\n",
       " 'Poecilotheria regalis',\n",
       " 'Psechrus sinensis',\n",
       " 'Steatoda grossa',\n",
       " 'Stegodyphus dumicola',\n",
       " 'Stegodyphus mimosarum',\n",
       " 'Tetragnatha kauaiensis',\n",
       " 'Tetragnatha versicolor',\n",
       " 'Trichonephila antipodiana',\n",
       " 'Trichonephila clavata',\n",
       " 'Trichonephila clavipes',\n",
       " 'Trichonephila inaurata madagascariensis',\n",
       " 'Trichonephila senegalensis',\n",
       " 'Uloborus diversus'}"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "species = set(silk_df[\"Species\"].to_list())\n",
    "species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "e3868472",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_aa_comp(seq):\n",
    "    total = len(seq)\n",
    "    aa_count = Counter(seq).most_common(5)\n",
    "    aa_comp = {k : round(v/total * 100,2)  for (k,v) in aa_count}\n",
    "    return aa_comp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "62c7f6f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "silk_df[\"Amino acid composition\"] = silk_df[\"Sequence\"].apply(get_aa_comp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "36d4def5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    {'S': 9.47, 'E': 7.69, 'L': 7.1, 'P': 7.1, 'R'...\n",
       "1    {'G': 40.38, 'S': 20.53, 'A': 15.29, 'Y': 5.2,...\n",
       "2    {'G': 47.6, 'A': 23.75, 'Q': 9.51, 'S': 5.9, '...\n",
       "3    {'G': 35.49, 'A': 19.01, 'P': 14.93, 'S': 11.4...\n",
       "4    {'G': 39.51, 'A': 17.97, 'P': 16.58, 'S': 9.02...\n",
       "                           ...                        \n",
       "3    {'A': 22.59, 'S': 10.16, 'E': 10.11, 'L': 9.89...\n",
       "4    {'A': 21.19, 'E': 9.36, 'S': 8.99, 'L': 8.49, ...\n",
       "5    {'A': 15.31, 'S': 12.27, 'L': 10.34, 'E': 9.33...\n",
       "6    {'A': 22.71, 'S': 9.37, 'L': 9.22, 'E': 9.22, ...\n",
       "7    {'A': 14.31, 'S': 12.91, 'L': 10.42, 'E': 9.95...\n",
       "Name: Amino acid composition, Length: 1001, dtype: object"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "silk_df[\"Amino acid composition\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "d6609b0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "silk_df.to_csv(\"silk_data.csv\", index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
